lib/msf/ui/console/command_dispatcher/core.rb

```diff
@@ -2128,71 +2166,100 @@ def cmd_getg_tabs(str, words)
   #
   def cmd_unset(*args)
+    if args.include?('-h') || args.include?('--help')
+      cmd_unset_help
+      return
+    end
 
     # Figure out if these are global variables
     global = false
 
-    if (args[0] == '-g')
-      args.shift
-      global = true
+    @@unset_opts.parse(args) do |opt, idx, val|
+      case opt
+      when '-g'
+        global = true
+      end
+    end
+
+    variable_names = args.reject { |arg| arg.start_with?('-') }
+
+    # No variable names? No cookie.
+    if variable_names.empty?
+      cmd_unset_help
+      return false
     end
 
     # Determine which data store we're operating on
-    if (active_module and global == false)
+    if active_module && !global
       datastore = active_module.datastore
     else
       datastore = framework.datastore
     end
 
-    # No arguments?  No cookie.
-    if (args.length == 0)
-      cmd_unset_help
-      return false
+    is_all_variables = variable_names[0] == 'all'
+    if is_all_variables
+      variable_names = datastore.keys
+      variable_names += Msf::DataStore::GLOBAL_KEYS if global
+      variable_names += ['PAYLOAD'] if !global && active_module && (active_module.exploit? || active_module.evasion?)
+      variable_names = variable_names.uniq(&:downcase)
     end
 
-    # If all was specified, then flush all of the entries
-    if args[0] == 'all'
-      print_line("Flushing datastore...")
+    print_line("Unsetting datastore...") if is_all_variables
 
-      # Re-import default options into the module's datastore
-      if (active_module and global == false)
-        active_module.import_defaults
-      # Or simply clear the global datastore
-      else
-        datastore.clear
-      end
-
-      return true
-    end
-
-    while ((val = args.shift))
-      if (driver.on_variable_unset(global, val) == false)
-        print_error("The variable #{val} cannot be unset at this time.")
+    variable_names.each do |variable_name|
+      if driver.on_variable_unset(global, variable_name) == false
+        print_error("The variable #{variable_name} cannot be unset at this time.") # unless variable_name.casecmp?('PAYLOAD')
         next
       end
 
-      print_line("Unsetting #{val}...")
-
-      datastore.delete(val)
+      print_line("Unsetting #{variable_name}...") unless is_all_variables
+      datastore.unset(variable_name)
+    end
+
+    # Do a final pass over the datastore. If a user has unset a variable - but it continues to have a value either through
+    # option defaults, or being globally set it might be confusing to users. In this scenario, log out a helpful message.
+    #
+    # i.e. the scenario of a user unsetting 'RHOSTS', but the value continues to inherit from the global framework datastore.
+    unless is_all_variables
+      variable_names.each do |variable_name|
+        search_result = datastore.search_for(variable_name)
+        if search_result.fallback?
```

Comment:
With the feature flag does this need a guard?
```suggestion
        if search_result.respond_to(:fallback?) && search_result.fallback?
```

Reply:
I've added a new `cmd_unset_with_fallbacks` method for this logic in its entirety now :+1: